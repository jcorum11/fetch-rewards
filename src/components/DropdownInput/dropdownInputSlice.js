import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'
import dropdownApi from '../../api/dropdown'

export const fetchOptions = createAsyncThunk('dropdown/fetchOptionsStatus', async () => {
  const response = await dropdownApi.fetchOptions()
  return response
})

const dropdownSlice = createSlice({
  name: 'dropdown',
  initialState: {
    options: [],
    status: 'idle',
    selectedState: '',
    occupation: ''
  },
  reducers: {
    setState: (state, action) => {
      state.selectedState = action.payload
    },
    setOccupation: (state, action) => {
      state.occupation = action.payload
    }
  },
  extraReducers: builder => {
    builder.addCase(fetchOptions.pending, (state, action) => {
      state.status = 'loading'
    })
    builder.addCase(fetchOptions.fulfilled, (state, action) => {
      state.status = 'complete'
      state.options = action.payload
    })
  }
})

export const selectOptions = state => state.dropdown.options
export const selectOptionsFetchStatus = state => state.dropdown.status
export const selectOccupation = state => state.dropdown.occupation
export const selectState = state => state.dropdown.selectedState

export const { setState, setOccupation } = dropdownSlice.actions

export default dropdownSlice.reducer
