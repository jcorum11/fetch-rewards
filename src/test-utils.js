import React from 'react'
import { render as rtlRender } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
// Import your own reducer
import dropdownReducer from './components/DropdownInput/dropdownInputSlice'
import textInputReducer from './components/TextInput/textInputSlice'
import appReducer from './pages/App/appSlice'

function render(
  ui,
  {
    preloadedState,
    store = configureStore({ reducer: { dropdown: dropdownReducer, textInput: textInputReducer, app: appReducer }, preloadedState }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

// re-export everything
export * from '@testing-library/react'
// override render method
export { render }