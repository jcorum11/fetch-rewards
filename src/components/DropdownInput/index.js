/** @jsxImportSource @emotion/react */
import { useDispatch } from 'react-redux'
import { useSelector } from 'react-redux'
import { selectOptions, setOccupation, setState} from './dropdownInputSlice'
import { css } from '@emotion/react';

const DropdownInput = ({ type }) => {
  const dispatch = useDispatch()
  const options = useSelector(selectOptions)
  const key = `${type}s`.toLowerCase()
  const handleChange = (event) => {
    if (type === 'State') {
      dispatch(setState(event.target.value))
    } else if (type === 'Occupation') {
      dispatch(setOccupation(event.target.value))
    }
  }
  return (
    <div css={inputContainer}>
      <label css={label} htmlFor={type} data-testid={`${type}input-label`.toLowerCase()}>{type}: </label>
      <select css={input} name={type.toLowerCase()} id={type.toLowerCase()} onChange={handleChange}>
      <option value=''>{type === 'State' ? 'Select a state' : 'Select an occupation'}</option>
        {
          options[key] && options[key].map((option, index) => {
            return <option value={type === 'State' ? option.name : option} key={`option-${index}`}>{type === 'State' ? option.name : option}</option>
          })
        }
      </select>
    </div>
  )
}

const label = css`
display: block;
color: white
`

const inputContainer = css`
padding: .5rem;
`

const input = css`
width: 100%;
height: 2rem;
border-radius: 5px;
`

export default DropdownInput
