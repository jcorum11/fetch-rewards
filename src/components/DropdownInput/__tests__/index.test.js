import React from 'react'
import DropdownInput from '..'
import { render, cleanup } from '../../../test-utils'
import '@testing-library/jest-dom/extend-expect'

afterEach(cleanup)

describe('DropdownInput', () => {
  it('Renders', async () => {
    render(<DropdownInput type='State' />)
  })

  it('Matches snapshot', async () => {
    const { asFragment } = render(<DropdownInput type='State' />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('Inserts text into label', () => {
    const { getByTestId } = render(<DropdownInput type='State' />)
    expect(getByTestId('stateinput-label')).toHaveTextContent('State:')
  })
})