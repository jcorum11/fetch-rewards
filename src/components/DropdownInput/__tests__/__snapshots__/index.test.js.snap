// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownInput Matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-60wqy5-DropdownInput"
  >
    <label
      class="css-1iyxaku-DropdownInput"
      data-testid="stateinput-label"
      for="State"
    >
      State: 
    </label>
    <select
      class="css-vi962m-DropdownInput"
      id="state"
      name="state"
    >
      <option
        value=""
      >
        Select a state
      </option>
    </select>
  </div>
</DocumentFragment>
`;
